README.md
# KJDSS

The KJ Digital Signal Synchronizer (KJDSS) library allows developers of Java Sound based applications to implement real time digital signal processors like scopes and spectrum analysers. It includes an all Java scope and spectrum analyser component.

These package is open source and should be used at your own risk. 
Please read the license agreement. (GNU_LGP_License.htm)

This software package is available for download at: 

    http://www.fudcom.com/

If you have any questions, problems, or find any bugs, please email me at: 

    info@fudcom.com

----

## What is KJ Digital Signal Synchronizer (KJDSS)?

The KJ Digital Signal Synchronizer (renamed from KJ DSP Package) is a library for Java Sound that can be used to synchronize visual effects to speaker output.

## What is new in this version?

See the 'ReleaseNotes.txt' for details. 
 
## How do I use it?

You can look at the 'KJDSSSample.java' source file included in the 'kjdss130_src.zip' file OR consider this pseudo code example:

    KJDigitalSignalSynchronizer dss = new KJDigitalSignalSynchronizer();

    dss.start( sourceDataLine );

    while( decoder.hasMoreFrames() ) {

        byte[] wAudioDataFrame = decoder.decodeFrame();

        dss.write( wAudioDataFrame );

    }

- You must call the 'start' method on the DSS with the 'sourceDataLine' that will be sending output to the speakers. The DSS class will synchronize the output using the getLongFramePosition() of the sourceDataLine.
- Next you must write the audio data to the DSS which will also write the data to the source data line. The DSS will fire processing events to any KJDigitalSignalProcessor's registered with it.
- You must then register a KJDigitalSignalProcessor to the DSS audio data consumer that will be listening for processing events.
- Finally, you need process the audio data in some way to render it to the display.

## The KJ DSS Scope/Spectrum Analyzer Component

There is an AWT component called KJScopeAndSpectrumAnalyzer (com.fudcom.kjdss.ui) provided with this package that uses the KJ DSS.


 


release/ReleaseNotes.txt
KJ Digital Signal Synchronizer
------------------------------

These package is open source and should be used at your own risk. 
Please read the license agreement. (GNU_LGP_License.htm)

This software is available for download at: 

    http://www.fudcom.com/

If you have any questions, problems, or find any bugs, please email me at: 

    info@fudcom.com

*********************************************************************************************

Release Notes KJDSS v1.3.2-SNAPSHOT (January 2017)
-----------------------------------

Imported to Github.

Release Notes KJDSS v1.3.0 (January 2010)
--------------------------

*** New Stuff and Changes

- Added 'calculateFrequencyTable' to KJFFT class to create a frequency table (in hertz) that correlates to the bands of the FFT output data.
- Improved 'BandDistribution' interface of KJScopeAndSpectrumAnalyser to be more closely coupled to the FFT instance. 
- The KJScopeAndSpectrumAnalyser band distribution table now includes the frequency range (in hertz) for each visible band of the spectrum analyser.
- Added customizable 'BandGain' interface to KJScopeAndSpectrumAnalyser which can apply different levels of gain to each band of FFT output data. 
- Improved look of and added options to scope, spectrum analyser and, volume meter. 
	- Spectrum analyser can now render the frequency of each band.
	- Spectrum analyser is using the new BandGain function to 'boost' the high frequencies for a more pleasing display.
	- Spectrum analyser bars are now spaced evenly no matter the size. 
	- Volume meter is more accurate and is channel aware. Displays 1 bar for mono, and 2 for stereo.
- KJScopeAndSpectrumAnalyser class is no longer supported by earlier versions of KJDSS (or KJDSP).

*** Fixes

- Data normaliser used by spectrum analyser supports more PCM raw audio data formats. Still some unsupported formats though (big-endian, unsigned)
- 8 bit audio data was trashed by the data normalizer due to a signed/unsigned conversion error. Caused incorrectly normalized audio data to be pass to the DSP components.


Release Notes KJDSS v1.2.2 (December 2009)
--------------------------

*** Fixes

- Fix a logic bug that caused clicks during playback when using KJDSS with source data line writing enabled. (Thanks Jeroen)


Release Notes KJDSS v1.2.1 (May 2009)
--------------------------

*** Changes

- Added functionality to modify the band distribution of the spectrum analyzer. See KJScopeAndSpectrumAnalyzer$BandDistribution class.
- Added option to prevent the KJDigitalSignalSynchronizer from writing to the source data line. See KJDigitalSignalSynchronizer.setSourceDateLineWriteEnabled().
- Default spectrum analyzer gain adjusted to 3.0
- KJScopeAndSpectrumAnalyzer class can now be easily extended and customized. Opened up many private methods and fields.

*** Fixes

- Fixed band distribution of sub-sonic bands. The first band more accurately displays bass drums and other low frequency sounds.
- Minor performance improvements.
- Sample program displays directories in the file chooser.


Release Notes KJDSS v1.2 (February 2009)
------------------------

*** Changes

- Package renamed to KJ Digital Signal Synchronizer (KJDSS) instead of KJ Digitial Signal Processor (KJ DSP)
- Java package hierarchy changes. ( kjdss )
- API changes to KJDigitalSignalProcessor, please update existing processors.
- Separated core KJ classes from KJ DSS and therefore KJDigitalSignalAudioDataConsumer is now renamed to KJDigitalSignalSynchronizer.
- Moved all spectrum analyser fine tuning code into the KJScopeAndSpectrumAnalyser (and out of KJFFT)
- Automatically calculate sample size for FPS and audio format.
- Provided direct access to DSS data buffer and other information from within a DSP via the KJDigitalSignalSynchronizer.Context object.
- Removed default sample processing into float[] channels.
- Added sample class to demo the KJDSS and KJScopeAndSpectrumAnalyser component

*** Fixes

- Fixed FFT calculation to prevent feed back at higher frequencies.
- Fixed FFT class to handle difference sample sizes properly.
- Fixed spectrum analyser band merging and averaging calculations. 5 band output looks roughly the same as 20 band or 120 band etc. )
- Applied Log band distribution to spectrum analyser for band merging. This spreads the lower frequency bands out wider for a nicer view of human voices in the spectrum.  
- Applied proper Log band gain spectrum analyser. Applies a Log multiplier to each band to "boost" the higher frequency bands, which generate a more uniform spectrum analyser output.
- Fixed GUI deadlocks with KJScopeAndSpectrumAnalyser component.
- Fixed buffer overflow problems when writing large amounts of audio data ( > 256kb ) to the DSS.


samples/src/main/java/KJDSSSample.java
/**
 * Project: KJ DSS
 * File   : KJDSSSample.java
 *
 * Author : Kristofer Fudalewski
 * Email  : sirk_sytes@hotmail.com   
 * Website: http://sirk.sytes.net
 */

import java.awt.BorderLayout;
import java.awt.Component;
import java.awt.HeadlessException;
import java.io.File;

import javax.sound.sampled.AudioFileFormat;
import javax.sound.sampled.AudioInputStream;
import javax.sound.sampled.AudioSystem;
import javax.sound.sampled.SourceDataLine;
import javax.swing.JFileChooser;
import javax.swing.JFrame;
import javax.swing.filechooser.FileFilter;

import com.fudcom.kjdss.KJDigitalSignalSynchronizer;
import com.fudcom.kjdss.ui.KJScopeAndSpectrumAnalyzer;

/**
 * @author Kris Fudalewski
 *
 * Sample application demonstrating KJDSS 
 */
public class KJDSSSample extends JFrame {

	private static final int READ_BUFFER_SIZE = 1024 * 4;
	
	private KJDigitalSignalSynchronizer dss;
	
	public KJDSSSample() throws HeadlessException {
		
		super();
		
		initGUI();
		
		initDSS();
		
        setVisible( true );
        
        mainLoop();

	}
    
	private void initDSS() {
		
		// -- Create a DSS.
		dss = new KJDigitalSignalSynchronizer();
		
		// -- Create DSP that comes with KJDSS (also used in KJ).
		KJScopeAndSpectrumAnalyzer wDsp = new KJScopeAndSpectrumAnalyzer();
		
		// -- Add DSP to DSS.
		dss.add( wDsp );
		
		// -- Add DSP as component to JFrame
        add( (Component)wDsp );
		
	}
	
	private void initGUI() {
		
		setTitle( "KJDSS - Sample" );

		setLayout( new BorderLayout() );
		
        setSize( 456, 208 );
        
        setDefaultCloseOperation( JFrame.EXIT_ON_CLOSE );
        
	}
	
	private void mainLoop() {
		
		while( true ) {
		
			// -- Choose a WAV file.
			File wWavFile = chooseFile();

			// -- Play the chosen WAV file, otherwise exit.
			if ( wWavFile != null ) {
				playFile( wWavFile );
			} else {
				break;
			}
			
		}
		
		System.exit( 0 );
		
	}
	
	private void playFile( File pWavFile ) {
		
        try {
	        
        	// -- Load the WAV file.
            AudioInputStream wAs = AudioSystem.getAudioInputStream( pWavFile );
            
            // -- Create a source data line in the format of the file.
        	SourceDataLine wSdl = AudioSystem.getSourceDataLine( wAs.getFormat() );

        	// -- Open the source data line and start it.
        	wSdl.open();
        	wSdl.start();

        	// -- Have the DSS monitor the source data line.
            dss.start( wSdl );
        	
            // -- Allocate a read buffer.
        	byte[] wRb = new byte[ READ_BUFFER_SIZE ];
        	int    wRs = 0;
        	
        	// -- Read from WAV file and write to DSS (and the monitored source data line)
        	while( ( wRs = wAs.read( wRb ) ) != -1 ) {
        		dss.writeAudioData( wRb, 0, wRs );
        	}

        	// -- EOF, stop monitoring source data line.
        	dss.stop();
        	
        	// -- Stop and close the source data line.
        	wSdl.stop();
        	wSdl.close();
            
        } catch( Exception pEx ) {
        	pEx.printStackTrace();
        }
		
	}
	
	
	private File chooseFile() {
		
		JFileChooser wFc = new JFileChooser();
		
		wFc.setFileFilter( 
			new FileFilter() {

				@Override
				public boolean accept( File pFile ) {
					return pFile.getName().toLowerCase().endsWith( ".wav" ) || pFile.isDirectory();
				}

				@Override
				public String getDescription() {
					return "*.wav";
				}
				
			}
		);
		
		wFc.setMultiSelectionEnabled( false );
		
		wFc.showOpenDialog( this );
		
		return wFc.getSelectedFile();
		
	}
	
	
	public static void main( String[] pArgs ) {
		new KJDSSSample();
	}

}


samples/src/main/java/KJSampleBuffer.java


import javazoom.jl.decoder.Obuffer;

/**
 * @author Kris
 *
 * TODO To change the template for this generated type comment go to
 * Window - Preferences - Java - Code Generation - Code and Comments
 */
public class KJSampleBuffer extends Obuffer {

	  private byte[] buffer;
	  private int[]  bufferp;

	  private int    channels;
	  private int	 frequency;
	  
//	  float[] fft_real;
//	  double fft_imaginary[];
	  
//	  int    fft_pos;
						   
	  public KJSampleBuffer( int pSampleFrequency, int pNumberOfChannels ) {

          buffer = new byte[ OBUFFERSIZE * pNumberOfChannels ];
		  bufferp = new int[ MAXCHANNELS ];
	  	
	//	  fft_real = new float[ OBUFFERSIZE ];
	//	  fft_imaginary = new double[ OBUFFERSIZE ];
		  
		  channels = pNumberOfChannels;
		  frequency = pSampleFrequency;
		
	  }

	  public int getChannelCount() {
		return channels;  
	  }
	  
	  public int getSampleFrequency() {
		  return frequency;
	  }
	  
	  public byte[] getBuffer() {
	       return this.buffer;  
	  }
	  
	  public int getBufferLength() {
		  return bufferp[ 0 ];
	  }
	  
//	  public float[] getAudioData() {
//	  	return fft_real;
//	  }
//	  
//	  public float[] getFFTForSample() {
//	  	
//	      return KJFFT.fftMag( fft_real, 1024 );
//	  	
////	      return fft_real;
//	      
//	  }
	  
	  /**
	   * Takes a 16 Bit PCM sample.
	   */
	  public void append( int pChannel, short pValue ) {
	  	
 		  buffer[ bufferp[ pChannel ] ]     = (byte)pValue;
 		  buffer[ bufferp[ pChannel ] + 1 ] = (byte)( pValue >>> 8 );
 		  
		  bufferp[ pChannel ] += channels << 1;	  	
	      
	  }
	  
		public void appendSamples( int pChannel, float[] pSamples ) {
			
		    int wPosition = bufferp[ pChannel ];
			
		    for ( int i = 0; i < 32; i++ ) {
		    	
		    	float wFloatSample = pSamples[ i ];
		    	
//		    	if ( pChannel == 0 ) {
//		    		fft_real[ fft_pos + i ] = pSamples[ i ] / 32767.0f;
//		    	}
		    	
	     //       fft_imaginary[ fft_pos + i ] = 0.0;					  
		    	
				wFloatSample = ( wFloatSample > 32767.0f ? 32767.0f : ( wFloatSample < -32767.0f ? -32767.0f : wFloatSample ) );
				
				short wSample = (short)wFloatSample;
				
				buffer[ wPosition ]     = (byte)wSample;
				buffer[ wPosition + 1 ] = (byte)( wSample >>> 8 );
				
				wPosition += ( channels << 1 );
				
		    }
			
			bufferp[ pChannel ] = wPosition;
			
//			if ( pChannel == 0 ) {
//				fft_pos += 32;
//			}
			
		}
	  
	  
	  /**
	   * Write the samples to the file (Random Acces).
	   */
	  public void write_buffer(int val)
	  {
					  
		//for (int i = 0; i < channels; ++i) 
		//	bufferp[i] = (short)i;

	  }

	  public void close()
	  {}
	  
	  /**
	   *
	   */
	  public void clear_buffer() {
	  	
          for( int i = 0; i < channels; i++ ) {
              bufferp[ i ] = i << 1;
          }
          
      //   fft_pos = 0;
          
	  }
	  
	  /**
	   *
	   */
	  public void set_stop_flag()
	  {}

}




samples/src/main/java/KJDSSSampleApplet.java
/**
 * Project: KJ DSS
 * File   : KJDSSSampleApplet.java
 *
 * Author : Kristofer Fudalewski
 * Email  : sirk_sytes@hotmail.com   
 * Website: http://sirk.sytes.net
 */

import java.applet.Applet;
import java.awt.BorderLayout;
import java.awt.Color;
import java.awt.Component;
import java.awt.EventQueue;
import java.awt.HeadlessException;
import java.io.File;
import java.io.FileInputStream;

import javax.sound.sampled.AudioFormat;
import javax.sound.sampled.AudioInputStream;
import javax.sound.sampled.AudioSystem;
import javax.sound.sampled.SourceDataLine;
import javax.swing.JFileChooser;
import javax.swing.JOptionPane;
import javax.swing.filechooser.FileFilter;

import javazoom.jl.decoder.Bitstream;
import javazoom.jl.decoder.Decoder;
import javazoom.jl.decoder.Header;
import javazoom.jl.decoder.SampleBuffer;
import com.fudcom.kjdss.KJDigitalSignalSynchronizer;
import com.fudcom.kjdss.ui.KJScopeAndSpectrumAnalyzer;

/**
 * @author Kris Fudalewski
 *
 * Sample application demonstrating KJDSS 
 */
public class KJDSSSampleApplet extends Applet implements Runnable {

	private static final int READ_BUFFER_SIZE = 1024 * 4;
	
	private Thread  mainThread;
	private boolean active = true;
	
	private KJDigitalSignalSynchronizer dss;
	
	public KJDSSSampleApplet() throws HeadlessException {
		super();
	}
    
	private void initDSS() {
		
		// -- Create a DSS.
		dss = new KJDigitalSignalSynchronizer();
		
		// -- Create DSP that comes with KJDSS (also used in KJ).
		KJScopeAndSpectrumAnalyzer wDsp = new KJScopeAndSpectrumAnalyzer();
		
		wDsp.setSpectrumAnalyserBandCount( 128 );
		wDsp.setSpectrumAnalyserBandDistribution( 
			new KJScopeAndSpectrumAnalyzer.LogBandDistribution( 4, 4.0 ) );
		
		wDsp.setBackground( Color.black );
		wDsp.setForeground( Color.gray );
		wDsp.setDisplayModeToggleOnMouseClickEnabled( false );
		wDsp.setDisplayMode( KJScopeAndSpectrumAnalyzer.DISPLAY_MODE_SPECTRUM_ANALYSER );
		
		// -- Add DSP to DSS.
		dss.add( wDsp );
		
		// -- Add DSP as component to JFrame
        add( (Component)wDsp );
		
	}

	public void init() {
		
		setLayout( new BorderLayout() );
		
		initDSS();
		
	}

	public void start() {
		
        setVisible( true );
        
		active = true;
		
        mainThread = new Thread( this );
        mainThread.start();
		
	}
	
	public void stop() {

//		System.out.println( "STOP!" );
		
		active = false;
		
		try {
			mainThread.join();
		} catch( InterruptedException pEx ) {
			// -- Do nothing.
		}
		
	}
	
	public void destroy() {

//		System.out.println( "DESTRROY!" );

		dss = null;
		
	}
	
	public void run() {
		
		mainLoop();
		
	}
	
	private void mainLoop() {

		File wLastFolder = new File( "" );
		
		while( active ) {
		
			// -- Choose a WAV file.
			File wAudioFile = chooseFile( wLastFolder );

			// -- Play the chosen WAV file, otherwise exit.
			if ( wAudioFile != null ) {
				
				wLastFolder = wAudioFile.getParentFile();
				
				playFile( wAudioFile );
				
			} else {
				active = false;
			}
			
		}
		
	}
	
	private void playFile( File pFile ) {
		
		try {
		
			String wNfn = pFile.getName().toLowerCase();
			
			if ( wNfn.endsWith( ".mp3" ) ) {
				playMP3File( pFile );
			} else if ( wNfn.endsWith( ".wav" ) ) {
				playWAVFile( pFile );
			} else {

				EventQueue.invokeAndWait(
					new Runnable() {
						public void run() {
							JOptionPane.showMessageDialog( 
							KJDSSSampleApplet.this, 
							"Unsupported file format!", 
							"Play Error!", 
							JOptionPane.ERROR_MESSAGE );
						}
					}
				);
				
			}
			
		} catch( final Exception pEx ) {
			
			pEx.printStackTrace();
			
			try {
			
				EventQueue.invokeAndWait(
					new Runnable() {
						public void run() {
							JOptionPane.showMessageDialog( 
								KJDSSSampleApplet.this, 
								"An error occurred during playback:\r\n" + pEx.getMessage(), 
								"Play Error!", 
								JOptionPane.ERROR_MESSAGE );
						}
					}
				);
				
			} catch( Exception pMwex ) {
				// -- Do nothing.
			}
			
		}
		
	}
	
	private void playMP3File( File pMp3File ) throws Exception {
		
		FileInputStream wFis = new FileInputStream( pMp3File );
		
		Bitstream wBitstream = new Bitstream( wFis );
	
		Header wHeader = wBitstream.readFrame();
		
		Decoder wDecoder = new Decoder(); 

		wDecoder.setOutputBuffer( 
		    new KJSampleBuffer( 
		        wHeader.frequency(), 
				wHeader.mode() == Header.SINGLE_CHANNEL ? 1 : 2 ) );

		KJSampleBuffer wSampleBuffer = (KJSampleBuffer)wDecoder.decodeFrame( wHeader, wBitstream );
		
        // -- Create a source data line in the format of the file.
    	SourceDataLine wSdl = AudioSystem.getSourceDataLine( 
    		new AudioFormat( 
    			wDecoder.getOutputFrequency(),
    			wHeader.mode() == Header.SINGLE_CHANNEL ? 8 : 16,
    			wDecoder.getOutputChannels(),
    			true, 
    			false ) );

    	// -- Open the source data line and start it.
    	wSdl.open();
    	wSdl.start();

    	// -- Have the DSS monitor the source data line.
        dss.start( wSdl );
    	
        // -- Decode MP3 data while active.
		while( active && wHeader != null ) {
		
			try {
			
				// -- Send to audio device.
				dss.writeAudioData( wSampleBuffer.getBuffer(), 0, wSampleBuffer.getBufferLength() );
				
                wBitstream.closeFrame();
                
	            // -- Read next frame.
	    		wHeader = wBitstream.readFrame();
	    		
	    		if ( wHeader != null ) {
	    			wSampleBuffer = (KJSampleBuffer)wDecoder.decodeFrame( wHeader, wBitstream );
	    		}
				
			} catch( Exception pEx ) {
				//System.err.println( pEx );
				pEx.printStackTrace();
			}

		}
		
    	// -- EOF, stop monitoring source data line.
    	dss.stop();
    	
    	// -- Stop and close the source data line.
    	wSdl.stop();
    	wSdl.close();
		
    	wFis.close();
        	
	}
	
	private void playWAVFile( File pWavFile ) throws Exception {
		
    	// -- Load the WAV file.
        AudioInputStream wAs = AudioSystem.getAudioInputStream( pWavFile );
        
        // -- Create a source data line in the format of the file.
    	SourceDataLine wSdl = AudioSystem.getSourceDataLine( wAs.getFormat() );

    	// -- Open the source data line and start it.
    	wSdl.open();
    	wSdl.start();

    	// -- Have the DSS monitor the source data line.
        dss.start( wSdl );
    	
        // -- Allocate a read buffer.
    	byte[] wRb = new byte[ READ_BUFFER_SIZE ];
    	int    wRs = 0;
    	
    	// -- Read from WAV file and write to DSS (and the monitored source data line)
    	while( active && ( wRs = wAs.read( wRb ) ) != -1 ) {
    		dss.writeAudioData( wRb, 0, wRs );
    	}

    	// -- EOF, stop monitoring source data line.
    	dss.stop();
    	
    	// -- Stop and close the source data line.
    	wSdl.stop();
    	wSdl.close();
            
	}
	
	
	private File chooseFile( File pLastFolder ) {
		
//		return new File( "B:\\kfud\\downloads\\Imogen Heap - Discography\\Releases\\Imogen Heap - Speak For Yourself (2005)\\Imogen Heap - Speak For Yourself - 05 - Hide And Seek.mp3" );
//		return new File( "B:\\kfud\\downloads\\Nederlandse Top 40 week 49 (2009) NLT-Release\\09 - David Guetta Ft. Akon - Sexy Bitch.mp3" );
		
		final JFileChooser wFc = new JFileChooser( pLastFolder );
		
		wFc.setDialogTitle( "KJDSS Sample Applet - Choose an audio file:" );
		
		wFc.setFileFilter(
				
			new FileFilter() {

				@Override
				public boolean accept( File pFile ) {
					
					if ( pFile.isDirectory() ) {
						return true;
					}
					
					String wNfn = pFile.getName().toLowerCase();
					
					return wNfn.endsWith( ".wav" ) || wNfn.endsWith( ".mp3" );
				}

				@Override
				public String getDescription() {
					return "*.mp3, *.wav";
				}
				
			}
			
		);
		
		wFc.setMultiSelectionEnabled( false );
		
		try {
		
			EventQueue.invokeAndWait(
				new Runnable() {
					public void run() {
						wFc.showOpenDialog( KJDSSSampleApplet.this );	
					}
				}
			);
			
		} catch( Exception pEx ) {
			// -- Do nothing.
		}
		
		return wFc.getSelectedFile();
		
	}
	
	public static void main( String[] pArgs ) {
		new KJDSSSampleApplet();
	}

}


library/src/main/java/com/fudcom/kjdss/KJDigitalSignalProcessor.java
/**
 * Project: KJ DSS
 * File   : KJDigitalSignalProcessor.java
 *
 * Author : Kristofer Fudalewski
 * Email  : info@fudcom.com   
 * Website: http://www.fudcom.com
 */
package com.fudcom.kjdss;

import javax.sound.sampled.AudioFormat;


/**
 * @author Kris Fudalewski
 * 
 * Classes must implement this interface in order to be registered with the 
 * KJDigitalSignalSynchronizer class.
 *  
 */
public interface KJDigitalSignalProcessor {

	/**
	 * Called by the KJDigitalSignalSynchronizer during the call of the 'start' method.
	 * Allows a DSP to prepare any necessary buffers or objects according to the audio format of
	 * the source data line.
	 * 
	 * @param pSampleSize     The sample size that this DSP should be prepared to handle.
	 * @param pAudioFormat    The audio format of passed in through the start method.
	 */
	void initialize( int pSampleSize, AudioFormat pAudioFormat );
	
	/**
	 * Called by the KJDigitalSignalSynchronizer while the SourceDataLine is active.
	 * 
	 * @param pDssContext A context object containing a reference to the sample data to be processed
	 * 					  as well as other useful references during processing time.
	 */
    void process( KJDigitalSignalSynchronizer.Context pDssContext );
	
}


samples/src/main/java/AlSample.java


import java.awt.BorderLayout;
import java.awt.Component;
import java.awt.HeadlessException;
import java.io.File;

import javax.sound.sampled.AudioFormat;
import javax.sound.sampled.AudioSystem;
import javax.sound.sampled.DataLine;
import javax.sound.sampled.Line;
import javax.sound.sampled.Mixer;
import javax.sound.sampled.SourceDataLine;
import javax.sound.sampled.TargetDataLine;
import javax.swing.JFileChooser;
import javax.swing.JFrame;
import javax.swing.filechooser.FileFilter;

import com.fudcom.kjdss.KJDigitalSignalSynchronizer;
import com.fudcom.kjdss.ui.KJScopeAndSpectrumAnalyzer;

public class AlSample extends JFrame {

	// This string needs to be set to the first few characters needed to uniquely identify
	// the Mixer (by name) on which a TargetDataLine is to be opened for reading audio data.
	// Typical values: "Stereo", "CD", "Line"
	private final static String mixerNameStartString = "Mic";
																 
	private static final int READ_BUFFER_SIZE = 1024;
	
	private KJDigitalSignalSynchronizer dss;
	
	public AlSample() throws HeadlessException {
		
		super();
		
		initGUI();
		
		initDSS();
		
        setVisible( true );
        
        mainLoop();

	}
    
	private void initDSS() {
		
		// -- Create a DSS.
		dss = new KJDigitalSignalSynchronizer();
		
		// -- Create DSP that comes with KJDSS (also used in KJ).
		KJScopeAndSpectrumAnalyzer wDsp = new KJScopeAndSpectrumAnalyzer();
		
		// -- Add DSP to DSS.
		dss.add( wDsp );
		
		// -- Add DSP as component to JFrame
        add( (Component)wDsp );
		
	}
	
	private void initGUI() {
		
		setTitle( "KJDSS - Sample" );

		setLayout( new BorderLayout() );
		
        setSize( 456, 208 );
        
        setDefaultCloseOperation( JFrame.EXIT_ON_CLOSE );
        
	}
	
	private void mainLoop() {
		
//		while( true ) {
		
			// -- Choose a WAV file.
//			File wWavFile = chooseFile();
		
			// Instead of opening a file to read, get a TargetDataLine as an audio source
			// In this context it is a source of audio data
			TargetDataLine srcLine = null;
			// Same AudioFormat as used by .WAV files
			AudioFormat fmt = new AudioFormat(44100F,16,2,true,false);
			Line.Info reqLineInfo = new DataLine.Info(TargetDataLine.class,fmt);
			if (!AudioSystem.isLineSupported(reqLineInfo)) {
				System.out.println("TDL not supported");
				System.exit(-1);
			}
			Mixer.Info[] mixInfoAr = AudioSystem.getMixerInfo();
			System.out.println("Found " + mixInfoAr.length + " Mixers. Looking for Mixer \"" +
								mixerNameStartString + "...\"");
			for ( Mixer.Info mixInfo : mixInfoAr ) {
				Mixer mix = AudioSystem.getMixer(mixInfo);
				
				System.out.println( "Mixer: " + mixInfo.getName() );
				
				if (mix.isLineSupported(reqLineInfo)) {
					try {

						if (mixInfo.getName().startsWith(mixerNameStartString)) {
							System.out.println("Found Mixer \"" + mixInfo.getName() + "\"");
							srcLine = (TargetDataLine)(mix.getLine(reqLineInfo));
							System.out.println("Got TargetDataLine");
							srcLine.open(fmt);
							System.out.println("Opened TargetDataLine");
							srcLine.start();
							break;
						}
					} catch (Exception ex) {
						System.out.println("Can't get TargetDataLine on Mixer: " + ex.getMessage());
					}
				}
			}

			// This will never return...
			playStream(srcLine);			

			// -- Play the chosen WAV file, otherwise exit.
/*			if ( wWavFile != null ) {
				playFile( wWavFile );
			} else {
				break;
			}
*/			
//		}
		
		System.exit( 0 );
		
	}
	
	// was: private void playFile( File pWavFile ) {
	private void playStream( TargetDataLine tdl) {
		
        try {
        	
        	// No need for that AudioInputStream - can read from the TargetDataLine directly	        
        	// -- Load the WAV file.
//            AudioInputStream wAs = AudioSystem.getAudioInputStream( pWavFile );

            // -- Create a source data line in the same format as the TargetDataLine
 //       	SourceDataLine wSdl = AudioSystem.getSourceDataLine( tdl.getFormat() ); //wAs.getFormat() );

        	// -- Open the source data line and start it.
//        	wSdl.open();
//        	wSdl.start();

        	// Disable writing to the SourceDataLine, since some other appliation will already
        	// be playing the audio we wish to visualize
        	dss.setSourceDataLineWriteEnabled( false );
        	
        	// -- Have the DSS monitor the target data line.
            dss.start( tdl );
            
            // -- Allocate a read buffer.
        	byte[] wRb = new byte[ 1024 ];
        	int    wRs = 0;

        	// -- Read from WAV file and write to DSS (and the monitored source data line)
        	//while( ( wRs = wAs.read( wRb ) ) != -1 ) {
        	while( ( wRs = tdl.read( wRb, 0, wRb.length ) ) != -1 ) {
//        		
//		//		System.out.println( tdl.getLongFramePosition() );
//
        		dss.writeAudioData( wRb, 0, wRs );
        		
        	}

        	// The while loop above will never terminate
        	
        	// -- EOF, stop monitoring source data line.
        	dss.stop();
        	
        	// -- Stop and close the source data line.
//        	wSdl.stop();
//        	wSdl.close();
            
        } catch( Exception pEx ) {
        	pEx.printStackTrace();
        }
		
	}
	
	// Not used
	private File chooseFile() {
		
		JFileChooser wFc = new JFileChooser();
		
		wFc.setFileFilter( 
			new FileFilter() {

				@Override
				public boolean accept( File pFile ) {
					return pFile.getName().toLowerCase().endsWith( ".wav" ) || pFile.isDirectory();
				}

				@Override
				public String getDescription() {
					return "*.wav";
				}
				
			}
		);
		
		wFc.setMultiSelectionEnabled( false );
		
		wFc.showOpenDialog( this );
		
		return wFc.getSelectedFile();
		
	}
	
	
	public static void main( String[] pArgs ) {
		new AlSample();
	}

	public class Normalizer {

//		private float[] left;
//		private float[] right;

		private AudioFormat audioFormat;
		
		private float[][] channels;

		private int  sampleSize;
		private int  channelSize;
		private long audioSampleSize;
		
		public Normalizer( AudioFormat pFormat ) {

			audioFormat = pFormat;
			
			sampleSize = (int)( Math.round( audioFormat.getFrameRate() ) );
			
			channels = new float[ pFormat.getChannels() ][]; 
			
			for( int c = 0; c < pFormat.getChannels(); c++ ) {
				channels[ c ] = new float[ sampleSize ];
			}
			
			channelSize     = audioFormat.getFrameSize() / audioFormat.getChannels();
			audioSampleSize = ( 1 << ( audioFormat.getSampleSizeInBits() - 1 ) );
			
			//audioFormat.getEncoding()
			
		}
		
		public float[][] normalize( byte[] pData, int pPosition, int pLength ) {
			
			int wChannels  = audioFormat.getChannels();
			int wSsib      = audioFormat.getSampleSizeInBits();
			int wFrameSize = audioFormat.getFrameSize();
			
			// -- Loop through audio data.
			for( int sp = 0; sp < sampleSize; sp++ ) { 
				
				if ( pPosition >= pData.length ) {
					pPosition = 0;
				}
				
				int cdp = 0;
				
				// -- Loop through channels.
				for( int ch = 0; ch < wChannels; ch++ ) {

					// -- Sign least significant byte. (PCM_SIGNED)
					long sm = ( pData[ pPosition + cdp ] & 0xFF ) - 128;
//					long sm = ( pData[ pPosition + cdp ] ); // & 0xFF ) - 128;
					
					for( int bt = 8, bp = 1; bt < wSsib; bt += 8 ) {
						sm += pData[ pPosition + cdp + bp ] << bt;
						bp++;
					}
					
					// -- Store normalized data.
					channels[ ch ][ sp ] = (float)sm / audioSampleSize;
				
//					System.out.println( "SP[" + sp + "] " + channels[ ch ][ sp ] );
					
					cdp += channelSize;
					
				}
					
				pPosition += wFrameSize;
				
			}
			
//			System.out.println( "------------------------" );

			//	System.out.println( "FS: " + audioFormat.getFrameSize() );
			
//			if ( wCc == 1 ) {
//			
//				if ( wSs == 8 ) {
//					
//					for( int a = 0, c = pPosition; a < sampleSize; a++, c += wFs ) { 
//					
//						if ( c >= pData.length ) {
//							c = 0;
//						}
//						
//						channels[ 0 ][ a ] = (float)( ( ( pData[ c ] & 0xFF ) - 128 ) / 128.0f );
//						
////						System.out.print( channels[ 0 ][ a ]  + " " );
//						//right[ a ] = left[ a ];
//						
//					}
//					
//				} else if ( wSs == 16 ) {
//					
//					for( int a = 0, c = pPosition; a < sampleSize; a++, c += wFs ) { 
//						
//						if ( c >= pData.length ) {
//							c = 0;
//						}
//						
//						channels[ 0 ][ a ]  = (float)( ( (int)pData[ c + 1 ] << 8 ) + pData[ c ] ) / 32768.0f;
//						//right[ a ] = left[ a ];
//						
//	//					System.out.print( channels[ 0 ][ a ] + " " );
//						
//					}
//					
//				}
//				
//			} else if ( wCc == 2 ) {
//			
//				if ( wSs == 8 ) {
//				
//					for( int a = 0, c = pPosition; a < sampleSize; a++, c += wFs ) { 
//						
//						if ( c >= pData.length ) {
//							c = 0;
//						}
//						
//						channels[ 0 ][ a ] = (float)( ( ( pData[ c ] & 0xFF ) - 128 ) / 128.0f );
//						channels[ 1 ][ a ] = (float)( ( ( pData[ c + 1 ] & 0xFF ) - 128 ) / 128.0f );
//						
//					}
//					
//				} else if ( wSs == 16 ) {
//				
//					for( int a = 0, c = pPosition; a < sampleSize; a++, c += wFs ) { 
//						
//						if ( c >= pData.length ) {
//							c = 0;
//						}
//						
//						channels[ 0 ][ a ] = (float)( ( (int)pData[ c + 1 ] << 8 ) + pData[ c ] ) / 32768.0f;
//						channels[ 1 ][ a ] = (float)( ( (int)pData[ c + 3 ] << 8 ) + pData[ c + 2 ] ) / 32768.0f;
//						
//					}
//					
//				}
//				
//			}
			
			return channels;
			
		}
		
	}
	
}
